File: Algorithm/137. Single Number II.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-07-15 14:39:42
# @Last modified by:   WuLC
# @Last Modified time: 2016-07-15 14:39:50
# @Email: liangchaowu5@gmail.com

class Solution(object):
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        result = 0
        for i in xrange(32):
            bit = 0
            for num in nums:
                bit += num & (1 << i)
            bit %= 3
            if bit:
                if i == 31:  # if the highest bit is 1,it is negative 
                    result -= pow(2,31)
                else:
                    result += pow(2, i)
        return result

File: Algorithm/Java/137. Single Number II.java
/**
* Author: WuLC
* Date:   2017-05-06 16:50:47
* Last modified by:   WuLC
* Last Modified time: 2017-05-06 16:52:24
* Email: liangchaowu5@gmail.com
*/

// count the number of 1 at each bit then mod 3 
// time O(n), more specific O(32n)
public class Solution 
{
    public int singleNumber(int[] nums) 
    {
        int result = 0, bit;
        for(int i = 0; i<32; i++)
        {
            bit = 0;
            for (int num:nums)
                if ((num & (1<<i)) != 0)
                    bit++;
            bit %= 3;
            result ^= (bit<<i);
        }
        return result;
    }
}


// more concise solution
// referer: https://discuss.leetcode.com/topic/2031/challenge-me-thx/17
public class Solution 
{
    public int singleNumber(int[] nums) 
    {
        int ones = 0, twos = 0;
        for(int num:nums)
        {
            ones = ((ones ^ num) & ~twos);
            twos = ((twos ^ num) & ~ones);
        }
        return ones;
    }
}